filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,322 @@
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+		ACKEvents: func(events []interface{}) {
+			for _, event := range events {
+				if meta, ok := event.(eventMeta); ok {
+					meta.handler.ack(meta.message)
+				}
+			}
+		},
+		CloseRef: doneChannelContext(inputContext.Done),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "reading kafka input config")
+	}
+
+	saramaConfig, err := newSaramaConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing Sarama config")
+	}
+
+	input := &kafkaInput{
+		config:       config,
+		saramaConfig: saramaConfig,
+		context:      inputContext,
+		outlet:       out,
+		log:          logp.NewLogger("kafka input").With("hosts", config.Hosts),
+	}
+
+	return input, nil
+}
+
+func (input *kafkaInput) runConsumerGroup() {
+	// Sarama uses standard go contexts to control cancellation, so we need
+	// to wrap our input context channel in that interface.
+	context := doneChannelContext(input.context.Done)
+	handler := &groupHandler{
+		version: input.config.Version,
+		outlet:  input.outlet,
+	}
+
+	// Create a consumer group and make sure it's closed before we return.
+	consumerGroup, err :=
+		sarama.NewConsumerGroup(
+			input.config.Hosts, input.config.GroupID, input.saramaConfig)
+	if err != nil {
+		input.log.Errorw(
+			"Error initializing kafka consumer group", "error", err)
+		return
+	}
+	input.saramaWaitGroup.Add(1)
+	defer func() {
+		consumerGroup.Close()
+		input.saramaWaitGroup.Done()
+	}()
+
+	// Listen asynchronously to any errors during the consume process
+	go func() {
+		for err := range consumerGroup.Errors() {
+			input.log.Errorw("Error reading from kafka", "error", err)
+		}
+	}()
+
+	err = consumerGroup.Consume(context, input.config.Topics, handler)
+	if err != nil {
+		input.log.Errorw("Kafka consume error", "error", err)
+	}
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (input *kafkaInput) Run() {
+	input.runOnce.Do(func() {
+		go func() {
+
+			// If the consumer fails to connect, we use exponential backoff with
+			// jitter up to 8 * the initial backoff interval.
+			backoff := backoff.NewEqualJitterBackoff(
+				input.context.Done,
+				input.config.ConnectBackoff,
+				8*input.config.ConnectBackoff)
+			for {
+				// Try to start the consumer group event loop: create a consumer
+				// group client (wbich connects to the kafka cluster) and call
+				// Consume (which starts an asynchronous consumer).
+				input.runConsumerGroup()
+
+				// If runConsumerGroup returns, then either input.context.Done has
+				// been closed (in which case we should shut down) or there was an
+				// error, and we should try running it again after the backoff interval.
+				waitChan := make(chan struct{})
+				go func() {
+					defer close(waitChan)
+					backoff.Wait()
+				}()
+
+				select {
```

Comment:
you already initialized the backoff instance with the same done channel. Calling `Wait` should give you exactly the same behavior.